test/e2e/network/proxy.go

```diff
@@ -258,9 +262,124 @@ var _ = SIGDescribe("Proxy", func() {
 				framework.Failf(strings.Join(errs, "\n"))
 			}
 		})
+
+		ginkgo.It("A set of valid responses are returned for both pod and service ProxyWithPath", func() {
+
+			ns := f.Namespace.Name
+
+			framework.Logf("Creating pod...")
+			_, err := f.ClientSet.CoreV1().Pods(ns).Create(context.TODO(), &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "agnhost",
+					Labels: map[string]string{
+						"test": "response"},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Image:   imageutils.GetE2EImage(imageutils.Agnhost),
+						Name:    "agnhost",
+						Command: []string{"/agnhost", "porter"},
+						Env: []v1.EnvVar{{
+							Name:  "SERVE_PORT_80",
+							Value: "foo",
+						}},
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				}}, metav1.CreateOptions{})
+			framework.ExpectNoError(err, "failed to create pod")
+
+			err = wait.PollImmediate(podRetryPeriod, podRetryTimeout, checkPodStatus(f, "test=response"))
+			framework.ExpectNoError(err, "Pod didn't start within time out period")
+
+			framework.Logf("Creating service...")
+			_, err = f.ClientSet.CoreV1().Services(ns).Create(context.TODO(), &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-service",
+					Namespace: ns,
+					Labels: map[string]string{
+						"test": "response",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					Ports: []v1.ServicePort{{
+						Port:       80,
+						TargetPort: intstr.FromInt(80),
+						Protocol:   v1.ProtocolTCP,
+					}},
+					Selector: map[string]string{
+						"test": "response",
+					},
+				}}, metav1.CreateOptions{})
+			framework.ExpectNoError(err, "Failed to create the service")
+
+			transportCfg, err := f.ClientConfig().TransportConfig()
+			framework.ExpectNoError(err, "Error creating transportCfg")
+			restTransport, err := transport.New(transportCfg)
+			framework.ExpectNoError(err, "Error creating restTransport")
+
+			client := &http.Client{
+				CheckRedirect: func(req *http.Request, via []*http.Request) error {
+					return http.ErrUseLastResponse
+				},
+				Transport: restTransport,
+			}
+
+			// All methods for Pod ProxyWithPath return 200
+			httpVerbs := []string{"DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"}
+			for _, httpVerb := range httpVerbs {
```

Comment:
This test just touches API surfaces and checks the status codes are 200.
Is it fine to do that without checking actual API behaviors?
I mean POST API could create something and the corresponding test needs to check something is created in general. And the same checks should be implemented for the other APIs also to check API behaviors.

Reply:
The goal of the test is to confirm that the proxy is passing valid requests through and returns a valid responses back for the list of endpoints below. When planning this test the [feedback](https://app.slack.com/client/T09NY5SBT/C09QZ4DQB) was to run [porter](https://github.com/kubernetes/kubernetes/tree/master/test/images/agnhost/porter) from the standard agnhost test image.

```
                     endpoint                    |                          path                           |      kind
 ------------------------------------------------+---------------------------------------------------------+-----------------
  connectCoreV1PutNamespacedPodProxyWithPath     | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
  connectCoreV1PostNamespacedPodProxyWithPath    | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
  connectCoreV1PatchNamespacedPodProxyWithPath   | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
  connectCoreV1OptionsNamespacedPodProxyWithPath | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
  connectCoreV1HeadNamespacedPodProxyWithPath    | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
  connectCoreV1DeleteNamespacedPodProxyWithPath  | /api/v1/namespaces/{namespace}/pods/{name}/proxy/{path} | PodProxyOptions
```

```
                      endpoint                      |                            path                             |        kind
----------------------------------------------------|-------------------------------------------------------------|---------------------
 connectCoreV1PutNamespacedServiceProxyWithPath     | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
 connectCoreV1PostNamespacedServiceProxyWithPath    | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
 connectCoreV1PatchNamespacedServiceProxyWithPath   | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
 connectCoreV1OptionsNamespacedServiceProxyWithPath | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
 connectCoreV1HeadNamespacedServiceProxyWithPath    | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
 connectCoreV1DeleteNamespacedServiceProxyWithPath  | /api/v1/namespaces/{namespace}/services/{name}/proxy/{path} | ServiceProxyOptions
```

Reply:
> The goal of the test is to confirm that the proxy is passing valid requests through and returns a valid responses back for the list of endpoints below.

I am raising what is a valid response here.
Current test checks response code (HTTP200) only, doesn't the response body contain any values?

Reply:
Currently porter will respond with `foo` to any request in the response body. Updating the test to check for that as well.

Reply:
Is this verifying whether the correct verb is being received by porter? e.g. If I send a POST and it arrives at the container as a GET, that is incorrect.

Reply:
At the moment it isn't and I agree that shouldn't happen. Will consider how to update the test for this,

Reply:
I did check the above scenario when I was checking the base proxy endpoints for pods. I was just reflecting the response method back to the client.
```
import (
	"fmt"
	"github.com/gorilla/mux"
	"log"
	"net/http"
)

func responseFunction(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "responseFunction: %v -> %v", r.Method, r.RequestURI)
	fmt.Printf("responseFunction: \n%v -> %v\n", r.Method, r.RequestURI)
}

func handleRequests() {
	r := mux.NewRouter()
	r.HandleFunc("/", responseFunction).Methods("GET", "POST", "PUT", "DELETE", "PATCH")
	log.Fatal(http.ListenAndServe(":80", r))
}

func main() {
	handleRequests()
}
```
Should porter be updated to support this requirement? or should the test use another method?

Reply:
From conformance sub-project meeting today, @johnbelamaric suggested adding a flag option that would support this use case.

Current idea is that it would return a json payload that would include both the http method that porter received and the standard response for that port.